# Changelog

## [0.4] - 2019-02-13

### Added
- Poreplex now support multi-read FAST5 files as inputs.
- FAST5 files which had base-called with the ONT guppy can be used as inputs.

### Changed
- Fixed `filename` in `sequencing_summary.txt` points to wrong locations of
  FAST5 files when both FAST5 output and barcoding are turned on.
- Poly(A) dwell time measurement is now more robust to pepper and salt
  noises.
- Fixed BAM alignment writers which had generated broken files in
  multi-threaded processing.

## [0.3.1] - 2018-10-09

This minor release fixes an ABI compatibility issues in the binary packages.

## [0.3] - 2018-10-09

### Added
- Reads now can be demultiplexed without basecalling.
- With `--polya` turned on, poly(A) tail dwell time is written
  to `sequencing_summary.txt`. More detailed information is
  written to the dumped events by the `--dump-basecalled-events`
  switch.
- Shows detailed error messages in `poreplex.log`.
- The files generated by `--dump-basecalled-events` now includes
  the attributes for signal scaling parameters and the last position
  that DNA adapter ends within the signal.
- Extremely short sequences are now sorted into `fail`ed reads.
  The minimum required length can be changed with the
  `--minimum-length` switch. (Suggested by Nathan Roach)

### Changed
- Fixed a segmentation fault when using
  [albacore 2.3.3](https://community.nanoporetech.com/posts/albacore-2-3-3).
- Fixed an error that stops overall process by an invalid FAST5 file.
- `filename` in `sequencing_summary.txt` is now shown as a relative path
  from the output directory, not from the subdirectory for a read
  group.
- Fixed a problem that separate lines of FASTA, FASTQ or
  `sequencing-summary.txt` are mixed up in the output file sometimes.
- Turned off the chimeric read filter by default. Now the `--filter-chimera`
  option turns it back on.
- Updated the neural network model for barcode demultiplexing for
  even less false positives using randomly stitched signal fragments
  as a background.

